paddle/platform/profiler.cc

```diff
@@ -169,5 +172,149 @@ std::vector<std::vector<Event>> DisableProfiler() {
+          if (event_idx.find(event_name) == event_idx.end()) {
+            event_idx[event_name] = event_items.size();
+            EventItem event_item = {event_name, 1,          event_time,
+                                    event_time, event_time, event_time};
+            event_items.push_back(event_item);
+          } else {
+            int index = event_idx[event_name];
+            event_items[index].calls += 1;
+            // total time
+            event_items[index].total_time += event_time;
+            // min time
+            event_items[index].min_time =
+                std::min(event_time, event_items[index].min_time);
+            // max time
+            event_items[index].max_time =
+                std::max(event_time, event_items[index].max_time);
+          }
+          // remove the start marker from the list
+          pushed_events.erase((++rit).base());
+        } else {
+          std::cout << "Warning: can not find the start marker of event "
+                    << events[i][j].name();
+        }
+      }
+    }
+    // average time
+    for (auto& item : event_items) {
+      item.ave_time = item.total_time / item.calls;
+    }
+    // sort
+    if (sorted_by != EventSortingKey::kDefault) {
+      std::sort(event_items.begin(), event_items.end(),
+                [&](EventItem& a, EventItem& b) {
+                  switch (sorted_by) {
+                    case EventSortingKey::kCalls:
+                      return a.calls > b.calls;
+                    case EventSortingKey::kTotal:
+                      return a.total_time > b.total_time;
+                    case EventSortingKey::kMin:
+                      return a.min_time > b.min_time;
+                    case EventSortingKey::kMax:
+                      return a.max_time > b.max_time;
+                    default:
+                      return a.ave_time > b.ave_time;
+                  }
+                });
+    }
+    events_table.push_back(event_items);
+  }
+  // output events table
+  std::cout.setf(std::ios::left);
+  const int data_width = 12;
+  std::cout << std::setw(max_name_width + 4) << "Event" << std::setw(data_width)
+            << "Calls" << std::setw(data_width) << "Total"
+            << std::setw(data_width) << "Min." << std::setw(data_width)
+            << "Max." << std::setw(data_width) << "Ave." << std::endl;
+  for (size_t i = 0; i < events_table.size(); ++i) {
+    for (size_t j = 0; j < events_table[i].size(); ++j) {
+      EventItem& event_item = events_table[i][j];
+      std::cout << std::setw(max_name_width + 4) << event_item.name
+                << std::setw(data_width) << event_item.calls
+                << std::setw(data_width) << event_item.total_time
+                << std::setw(data_width) << event_item.min_time
+                << std::setw(data_width) << event_item.max_time
+                << std::setw(data_width) << event_item.ave_time << std::endl;
+    }
+  }
```

Comment:
Maybe the `ParseEvents` and print can be two functions.

Reply:
Done